Route53 provider configuration

simonkey007 opened the pull request.

simonkey007 commented:
@munnerz I already added RBAC roles that works for our cluster. Route53 DNS provider also has been successfully tested.

munnerz commented:
Hey @simonkey007 

Thanks for the work on this.

After chatting to some others on my team, I think we've decided that the `accessKeyID` and `hostedZoneID` are not classed as 'secret' values. I'm happy to be challenged on this if you disagree, I'm not by any means an expert in AWS authentication best practices.

Aside from that, the examples of how to use the route53 provider are much appreciated - I'm still working on writing proper docs with clearer examples, so this helps people get up and running for now.

If you could amend your commits accordingly and squash them down to one that'd be awesome - I'll get this merged in then!

allandegnan commented:
The following worked for me:

```
apiVersion: certmanager.k8s.io/v1alpha1
kind: Certificate
metadata:
  name: mytls
spec:
  secretName: mytls-tls
  issuerRef:
    name: letsencrypt-prod
    kind: ClusterIssuer
  dnsNames:
  - my.domain.com
  acme:
    config:
    - domains:
      - my.domain.com
      dns01:
        provider: route53
```

```
apiVersion: certmanager.k8s.io/v1alpha1
kind: ClusterIssuer
metadata:
  name: letsencrypt-prod
spec:
  acme:
    # The ACME server URL
    server: https://acme-v01.api.letsencrypt.org/directory
    # Email address used for ACME registration
    email: letsencrypt@my.email
    # Name of a secret used to store the ACME account private key
    privateKeySecretRef:
      name: letsencrypt-prod
    # Enable the HTTP-01 challenge provider
    http01: {}
    dns01:
      providers:
      - name: route53
        route53:
          accessKeyID: abcdefg
          region: eu-west-1
          secretAccessKeySecretRef:
            name: route53-config
            key: secret-access-key
```

It was quite brittle getting to that point though, but I'll reproduce and raise bugs specifically for those issues.

simonkey007 commented:
@munnerz I removed the changes that you marked as unnecessary.  I left `accessKeyID` and `hostedZoneID` variables as a string type and just added configuration for Route53 provider to the example files.

munnerz commented on docs/examples/acme-issuer.yaml:
```diff
@@ -27,8 +27,21 @@ spec:
+            name: route53-config
+            key: secret-access-key
+          # The Route53 hosted zone ID
+          hostedZoneID: DIKER8JPL21PSA
```
This field is optional, and only required if cert-manager cannot disambiguate between two different hosted zones for the same zone name. It may be worth noting that here?

Some more info here: https://github.com/jetstack-experimental/cert-manager/issues/94

munnerz commented on docs/examples/cluster-issuer.yaml:
```diff
@@ -0,0 +1,47 @@
+            name: route53-config
+            key: secret-access-key
+          # The Route53 hosted zone ID
+          hostedZoneID: DIKER8JPL21PSA
```
Same here - worth noting that this is an optional field and/or linking to the above mentioned issue 😄

munnerz commented on docs/api-types/issuer/spec.md:
```diff
@@ -127,6 +127,7 @@ clouddns:
 route53:
   accessKeyID: AKIAIOSFODNN7EXAMPLE
   region: eu-west-1
+  hostedZoneID: DIKER8JPL21PSA
```
Can we leave this bit out here? I'd rather not encourage people to set this field. It is likely to be removed pending discussion in https://github.com/jetstack-experimental/cert-manager/issues/94

munnerz commented:
/lgtm
/approve
